src/core/meshcop/border_agent.cpp

```diff
@@ -317,6 +346,227 @@ template <> void BorderAgent::HandleTmf<kUriRelayRx>(Coap::Message &aMessage, co
     FreeMessageOnError(message, error);
 }
 
+void BorderAgent::NotifyMeshCoPServiceChange(void)
+{
+    // We know all the TXT entries from OT core and the total size is less than 128.
+    static constexpr uint16_t kMaxTxtDataLen = 128;
+
+    uint8_t  txtDataBuffer[kMaxTxtDataLen];
+    uint16_t txtDataLen = kMaxTxtDataLen;
+
+    VerifyOrExit(mMeshCoPServiceChangedCallback.IsSet());
+    SuccessOrExit(mMeshCopTxtEncoder.EncodeTxtData(txtDataBuffer, txtDataLen));
+
+    mMeshCoPServiceChangedCallback.Invoke(txtDataBuffer, txtDataLen);
+
+exit:
+    return;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+// BorderAgent::MeshCopTxtEncoder
+
+Error BorderAgent::MeshCopTxtEncoder::EncodeTxtData(uint8_t *aOutput, uint16_t &aLength)
+{
+#if OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_1
+    static const char kThreadVersionString[] = "1.1.1";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_2
+    static const char kThreadVersionString[] = "1.2.0";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_3
+    static const char kThreadVersionString[] = "1.3.0";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_3_1
+    static const char kThreadVersionString[] = "1.3.1";
+#elif OPENTHREAD_CONFIG_THREAD_VERSION == OT_THREAD_VERSION_1_4
+    static const char kThreadVersionString[] = "1.4.0";
+#endif
+
+    Error                  error       = kErrorNone;
+    Message               *message     = Get<MessagePool>().Allocate(Message::kTypeOther);
+    const NetworkName     &networkName = Get<NetworkNameManager>().GetNetworkName();
+    const ExtendedPanId   &extPanId    = Get<ExtendedPanIdManager>().GetExtPanId();
+    const Mac::ExtAddress &extAddr     = Get<Mac::Mac>().GetExtAddress();
+    uint32_t               partitionId = Get<Mle::MleRouter>().GetLeaderData().GetPartitionId();
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+    Id id;
+#endif
+    StateBitmap   state;
+    int32_t       stateUint32;
+    Dns::TxtEntry txtEntry;
+
+    VerifyOrExit(message != nullptr, error = kErrorNoBufs);
+
+#if OPENTHREAD_CONFIG_BORDER_AGENT_ID_ENABLE
+    if (Get<BorderAgent>().GetId(id) == kErrorNone)
+    {
+        txtEntry.Init("id", id.mId, sizeof(id));
+        SuccessOrExit(error = txtEntry.AppendTo(*message));
+    }
+#endif
+    txtEntry.Init("nn", reinterpret_cast<const uint8_t *>(networkName.GetAsCString()),
```

Comment:
Can simplify this using `NameData` instead of `NetworkName`? (can use `NetworkName::GetAsNameData()`)

The `NameData` give the name to us as pointer to a byte buffer with its length.